### DISCUSSION TITLE
Make the session object more accessible

### DISCUSSION THREAD

**Discussion by falkoschindler** [zauberzeug/nicegui#732] [Ideas / Feature Requests]
In #373 we discussed how to "save variables between functions and requests of same session".
As shown in https://nicegui.io/documentation#sessions you can use the `request.session` dictionary. But this can be inconvenient, since you have to pass around the `request` object.
Therefore I want to open the discussion about a possible API that simplifies access to shared session state.

**Comment by amks1** [comment]
Is there a plan to implement this? I haven't been able to use Nicegui yet (🙁) because my use case depends heavily on sharing session info.

**Comment by rodja** [comment]
@amks1 we do not have a good idea on how the API should look like. Do you have any suggestions? What would be a convenient way for you to work with the session infos?

**Comment by amks1** [comment]
@rodja 
My prior experience is with PywebIO, as I had mentioned in https://github.com/zauberzeug/nicegui/discussions/373, so that kind of interface would be ideal. :)

**Reply by falkoschindler** [reply]
You're referring to [pywebio.session.local](https://pywebio.readthedocs.io/en/latest/session.html#pywebio.session.local), right?
So it's a "dictionary object" imported as `from pywebio.session import local`, which can be used as follows:

```py
def add():
    local.cnt = (local.cnt or 0) + 1

def show():
    ui.notify(local.cnt or 0)

ui.button('Add', add)
ui.button('Show', show)
```

**Reply by amks1** [reply]
> You're referring to [pywebio.session.local](https://pywebio.readthedocs.io/en/latest/session.html#pywebio.session.local), right? 

Yes, it's a dictionary variable in which I store mainly token details. Because my main app runs on Flask, I use that for all authentication and authorization, and redirect to the PywebIO server with a token in the URL. I wanted to do something similar for NiceGUI.

**Reply by rodja** [reply]
Intersting. The implementation with `ObjectDictProxy` and a lambda function is a neat trick. Worth trying for NiceGUI.

**Comment by rodja** [comment]
I have started a draft pull request for this feature in #967. Please have a look and tell me what you think.

**Comment by falkoschindler** [comment]
With #967 being released in 1.2.17 I'll close this discussion as completed. 🎉
